lib/include/modelSpec.h

```diff
@@ -436,6 +437,114 @@ class NNmodel
+                                     const typename CurrentSourceModel::ParamValues &paramValues,
+                                     const typename CurrentSourceModel::VarValues &varInitialisers,
+                                     int hostID = 0, int deviceID = 0)
+    {
+        if (!GeNNReady) {
+            gennError("You need to call initGeNN first.");
+        }
+        if (final) {
+            gennError("Trying to add a current source to a finalized model.");
+        }
+
+#ifdef MPI_ENABLE
+        // Determine the host ID
+        int mpiHostID = 0;
+        MPI_Comm_rank(MPI_COMM_WORLD, &mpiHostID);
+
+        // Pick map to add group to appropriately
+        auto &groupMap = (hostID == mpiHostID) ? m_LocalCurrentSources : m_RemoteCurrentSources;
+#else
+        // If MPI is disabled always add to local neuron groups and zero host id
+        auto &groupMap = m_LocalCurrentSources;
+        hostID = 0;
+#endif
+
+        // Add neuron group to map
+        auto result = groupMap.emplace(std::piecewise_construct,
+            std::forward_as_tuple(name),
+            std::forward_as_tuple(name, model,
+                                  paramValues.getValues(), varInitialisers.getInitialisers(), hostID, deviceID));
+
+        if(!result.second)
+        {
+            gennError("Cannot add a current source with duplicate name:" + name);
+            return NULL;
+        }
+        else
+        {
+            return &result.first->second;
+        }
+    }
+
+    //! Adds a new current source to the model using a singleton current source model created using standard DECLARE_MODEL and IMPLEMENT_MODEL macros
+    /*! \tparam CurrentSourceModel type of neuron model (derived from CurrentSourceModel::Base).
+        \param name string containing unique name of current source.
+        \param paramValues parameters for model wrapped in CurrentSourceModel::ParamValues object.
+        \param varInitialisers state variable initialiser snippets and parameters wrapped in CurrentSourceModel::VarValues object.
+        \return pointer to newly created CurrentSource */
+    template<typename CurrentSourceModel>
+    CurrentSource *addCurrentSource(const string &name,
+                                     const typename CurrentSourceModel::ParamValues &paramValues,
+                                     const typename CurrentSourceModel::VarValues &varInitialisers,
+                                     int hostID = 0, int deviceID = 0)
+    {
+        return addCurrentSource<CurrentSourceModel>(name, CurrentSourceModel::getInstance(), paramValues, varInitialisers, hostID, deviceID);
+    }
+
+    //! Injects an existing current source into an existing neuron group
+    /*! \param currentSourceName string name of the current source
+        \param targetNeuronGroupName string name of the targen neuron group */
+    void injectCurrent(const string &currentSourceName, const string &targetNeuronGroupName)
```

Comment:
I think a better API would be **just** to have an ``addCurrentSource`` method that takes the target neuron group name as a parameter. This API is weird because:

1. You can be left with 'orphan' current sources which don't inject anywhere
2. Especially once you add state variables, it's very ambiguous as to how a current source that injects into multiple, different sized populations works

Reply:
the reason I made separate addition and injection functions is the ability to inject a current source into multiple populations. The state vars of current sources could be renamed using both the name of the population that is injected and the name of the current source to disambiguate them. This does not solve the 'orphan' problem though

Reply:
Yeah - I think making one function simplifies a whole lot of things and removes a lot of edge cases. If you want to inject into multiple populations I don't see any issue with simply calling ``addCurrentSource`` multiple times.